### ISSUE TITLE
Race condition when multiple tasks concurrently start awaiting on a task just at it complete

### ISSUE THREAD

**Issue by ayende** [dotnet/coreclr#11265]
The following code starts a potentially expensive operation. That operation is cached, but because it can take a while, we want to cache the result, and if additional clients need the same operation, they can join the wait on the already started load.

The code below reproduce a hang on my machine, where out of five clients that started to listen to such a task, 4 of them are hanging.

The output of the code is something like:

```
0 (23, False, 827)
1 (23, False, 827)
2 (23, False, 827)
3 (23, False, 827)
4 (17, False, 1761016)
```

Note that 4 of those threads are still waiting on iteration 827, while one thread is executing freely.
However, all threads execute the tasks in the same order, so this shouldn't be the case.
I can verify that the task in question at this point has completed, but the waiting tasks hasn't been released.

I think that the race is in these two locations:

https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/Tasks/Task.cs#L4317

https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/Tasks/Task.cs#L3208

But the code seems to be handling this fine, so I'm not sure what is going on.

This seems to be a pretty stable race, and is easy to reproduce.

```c#
using System;
using System.Collections.Concurrent;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp1
{
    public class Program
    {
        private static readonly ConcurrentDictionary<int, Task<object>> _tasks = new ConcurrentDictionary<int, Task<object>>();
        private static readonly ConcurrentDictionary<int, object> _status = new ConcurrentDictionary<int, object>();

        private static void Main()
        {
            for (var i = 0; i < 5; i++)
            {
                _status[i] = -1;
                var index = i;
                Task.Run(() => SimulateCacheOfExpensiveOperations(index));
            }

            while (Console.KeyAvailable == false)
            {
                Console.Clear();
                foreach (var kvp in _status)
                    Console.WriteLine($"{kvp.Key} {kvp.Value}");
                Thread.Sleep(500);
            }
        }

        public static async Task SimulateCacheOfExpensiveOperations(int id)
        {
            var random = new Random(12);
            var i = 0;
            while (true)
            {
                var next = random.Next(0, 128);
                var task = new Task<object>(SimulateExpensiveOperation);

                var item = _tasks.GetOrAdd(next, task);
                if (item == task)
                    task.Start();

                _status[id] = (next, item == task, i++);
                await item;
            }
        }

        public static object SimulateExpensiveOperation()
        {
            Thread.Sleep(1);
            return "hi";
        }
    }
}
```

**Comment by stephentoub** [comment]
Are you sure it's a race in the implementation of Task and not just that `await` uses synchronous continuations by default?  If multiple synchronous continuations are hooked up to the same task that's not yet completed, then when it completes it's going to invoke those continuations synchronously as part of its completion, and if there are multiple of them, they'll get serialized... if the first one then runs for a long time, the rest won't have their continuations run until it's done.  What happens in your repro if you change:
```C#
new Task<object>(SimulateExpensiveOperation)
```
to
```C#
new Task<object>(SimulateExpensiveOperation, TaskCreationOptions.RunContinuationsAsynchronously)
```
?

**Comment by benaadams** [comment]
They all continue to count up
```
0 (53, False, 135783730)
1 (33, False, 142765657)
2 (91, False, 145137213)
3 (107, False, 131049528)
4 (28, False, 144458636)
```

**Comment by ayende** [comment]
@stephentoub Looks like you are correct. This looks like the underlying reason.
At least, adding this value resolve the issue in my small test. I'll test it to see if this happens the same way in my largest scope project.

I don't _think_ that I have anything there that is running for a long time without waiting for it.

**Comment by karelz** [comment]
Looks like it is resolved as problem in test code. I'll close the issue.
Feel free to reopen if you find more evidence there is CoreCLR bug.
